### PR TITLE
HPCC-19059 Allow all rows for a heap to be released in a single call

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman please review

File: ecl/hqlcpp/hqlcppds.cpp [hunk 1]
```diff
@@ -3031,6 +3031,7 @@ class EclccEngineRowAllocator : public CInterfaceOf<IEngineRowAllocator>
     virtual IOutputRowDeserializer *createInternalDeserializer(ICodeContext *ctx) override { throwUnexpected(); }
     virtual IEngineRowAllocator *createChildRowAllocator(const RtlTypeInfo *type) override { throwUnexpected(); }
     virtual void gatherStats(CRuntimeStatisticCollection & stats) override {}
+    virtual void releaseAllRows() override {}
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlcppds.cpp]: Should this be a throwUnexpected() ?

File: roxie/ccd/ccdserver.cpp [hunk 2]
```diff
@@ -7611,7 +7611,12 @@ class CRoxieServerHashDedupActivity : public CRoxieServerActivity
         {
             bestCompare=helper.queryCompareBest();
         }
-        virtual ~HashDedupTable() { _releaseAll(); }
+        virtual ~HashDedupTable()
+        {
+            //elementRowAllocator is a unique allocator, so all rows can be freed in a single call
+            elementRowAllocator->releaseAllRows();
+            tablecount =0;
```
Comment (richardkchapman) [inline roxie/ccd/ccdserver.cpp]: Missing space after =

**Comment by richardkchapman** [comment]
Should there also be a change in thorcommon.cpp ?

**Comment by richardkchapman** [comment]
@ghalliday

**Comment by ghalliday** [comment]
@richardkchapman pushed changes.  I didn't change aggregate because it generally frees items as it streams them out (and in general all rows are read).  There would be a minor advantage of releasing all in one go at the end, but the impact was fairly minimal on performance.

**Pull request merged**